### PR TITLE
Allow resetting VirtualizedDynamicScrollRectList (#540)

### PR DESCRIPTION
**Pull request by ahmed-shariff**
In response to #540 

This PR add a `Reset` public method to `VirtualizedScrollRectList`. The `OnValidate` method already contains the resetting functionality, this PR simply moves that to a new `Reset` method accessible from outside the class.

### CODE REVIEW

File: org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs
Comment (AMollis) [inline org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs:532]: Reset() is a MonoBehaviour editor mode event function.   Can you please rename this metho to avoid confusion.   Perhaps `ResetLayout`?

File: org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs
Comment (AMollis) [inline org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs:390]: See other comment....please update this comment once the Reset funciton is renamed

**Review by AMollis — CHANGES_REQUESTED** [review]
Please add a unit test that validates the "reset" functionality.

**Comment by ahmed-shariff** [comment]
Thank you for reviewing the PR. I could use some insight into how the tests are structured. The `VirtualizedScrollRectList` has a "test" in the dev template: https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/blob/f669f5aef4141ed3f949a2913a40e386c5444b2e/UnityProjects/MRTKDevTemplate/Assets/Scripts/VirtualizedScrollRectListTester.cs#L14

This has some of the functionality I would use in the unit test. Would it be ok to move that to the package? or should I leave that script as it is?

**Comment by AMollis** [comment]
> Thank you for reviewing the PR. I could use some insight into how the tests are structured. The `VirtualizedScrollRectList` has a "test" in the dev template:
> 
> https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/blob/f669f5aef4141ed3f949a2913a40e386c5444b2e/UnityProjects/MRTKDevTemplate/Assets/Scripts/VirtualizedScrollRectListTester.cs#L14
> 
> This has some of the functionality I would use in the unit test. Would it be ok to move that to the package? or should I leave that script as it is?

@ahmed-shariff , it looks like the original owner did not write Unit Tests for `VirtualizedDynamicScrollRectList`, which is unfortunate.  Given this, I'll be okay with one of the follow:

1. Setup a new unit test class here https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/tree/main/org.mixedrealitytoolkit.uxcore/Tests/Runtime and write a test to validate the reset functionality.
2. Update the Sample scene, [VirtualizedScrollRectList](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/blob/main/UnityProjects/MRTKDevTemplate/Assets/Scenes/Experimental/VirtualizedScrollRectList.unity), to demonstrate the "resetting" functionality.  Maybe add a button to resize and reset the list? 

I worry since this is not common/obvious functionality, there's a higher risk for regression if we don't provide a method to validate.

File: org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs
Comment (ahmed-shariff) [reply org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs]: It is now called reset layout

File: org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs
Comment (AMollis) [inline org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs:284]: I think we should check for a value change here.

```cs
    if (onInvisible != value)
    {
        onInvisible = value;
        // If this is changeing, it means how the items are populated is changing, hence reset.
        ResetLayout();
    }
```

File: org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs
Comment (AMollis) [inline org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs:265]: check for value change

File: org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs
Comment (AMollis) [inline org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs:206]: check for value change

**Review by AMollis — CHANGES_REQUESTED** [review]
Thank you so much for writing those Unit Test :)   Thank you !

I have some very minor suggestions....

File: org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs
Comment (ahmed-shariff) [reply org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs]: That's a good catch. Made the updates & added that to the tests as well.

**Review by AMollis — DISMISSED** [review]

**Comment by AMollis** [comment]
@shaynie , @marlenaklein-msft , @keveleigh .  

Once the AZP build completes, I'll merge unless you have feedback

File: org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs
Comment (keveleigh) [inline org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs:598]: I don't see where `resetCalled` is set to `true`. Should it be below here, right before the `StartCoroutine`?

**Review by keveleigh — CHANGES_REQUESTED** [review]

File: org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs
Comment (ahmed-shariff) [reply org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs:598]: Thank you for catching that. :D

File: org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs
Comment (keveleigh) [inline org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs:529]: I think this _may_ cause an issue with the batch mode test runner on our PR validation 😞 

https://docs.unity3d.com/ScriptReference/WaitForEndOfFrame.html

> Note: This coroutine is not invoked on the editor in batch mode.

Seems like there's a suggestion on to use `yield return Application.isBatchMode ? null : WaitForEndOfFrame();` at https://forum.unity.com/threads/do-not-use-waitforendofframe.883648/#post-5807410

Comment (keveleigh) [reply org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs:529]: I'd also be curious to know how this works in-editor when play mode isn't running, since this code path is also triggered by `OnValidate()`

**Review by marlenaklein-msft — DISMISSED** [review]

**Review by AMollis — CHANGES_REQUESTED** [review]
Changing my vote.... the Unit Tests seem to be failing.  Please fix Unit Tests

```
MixedReality.Toolkit.UX.Runtime.Tests.VirtualizedScrollRectListTests.TestVirtualizedScrollRectList_ResetLayout
  Non of the expected items were found in the scollable list (set1).
  Expected: True
  But was:  False
```

File: org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs
Comment (ahmed-shariff) [reply org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs]: I did not know about that. Thank you for pointing that out, pushing with that change. I was building on the code that was already there. In retrospect, `yield return null` would have the same outcome.

As for the OnValidate code path, from what I understand, it is what allows the UI to change when a value on the inspector is updated. There's probably a more optimal way to handle that without completely resetting it, but I don't want to touch that right now 😅

**Comment by ahmed-shariff** [comment]
> Changing my vote.... the Unit Tests seem to be failing. Please fix Unit Tests
> 
> ```
> MixedReality.Toolkit.UX.Runtime.Tests.VirtualizedScrollRectListTests.TestVirtualizedScrollRectList_ResetLayout
>   Non of the expected items were found in the scollable list (set1).
>   Expected: True
>   But was:  False
> ```

Not sure what is causing that. The tests were passing on my end. I have made a few updates to when frames are skipped, can we test if they pass again?

**Comment by AMollis** [comment]
> > Changing my vote.... the Unit Tests seem to be failing. Please fix Unit Tests
> > ```
> > MixedReality.Toolkit.UX.Runtime.Tests.VirtualizedScrollRectListTests.TestVirtualizedScrollRectList_ResetLayout
> >   Non of the expected items were found in the scollable list (set1).
> >   Expected: True
> >   But was:  False
> > ```
> 
> Not sure what is causing that. The tests were passing on my end. I have made a few updates to when frames are skipped, can we test if they pass again?

I'm guessing the Unit Tests were failing because of the previous `WaitForEndOfFrame()` calls which aren't supported in batch mode.  The CI pipeline is running the Unit Tests in batch mode.

File: org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs
Comment (AMollis) [reply org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs]: @keveleigh are you ok with this change?

File: org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs
Comment (AMollis) [inline org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs:602]: nit. add a new line after the `{`

**Review by AMollis — APPROVED** [review]
I approve.   Once @keveleigh approve we will merge

File: org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs
Comment (keveleigh) [inline org.mixedrealitytoolkit.uxcore/Experimental/List/VirtualizedScrollRectList.cs:293]:
```suggestion
                    // If this is changing, it means how the items are populated is changing, hence reset.
```

**Review by keveleigh — APPROVED** [review]

**Review by keveleigh — APPROVED** [review]

**Review by AMollis — APPROVED** [review]

**Review by AMollis — APPROVED** [review]

**Pull request merged**
